Prevent User from Sending Blank PMs

rdhammond opened the pull request:
Addressing issue #4872. Website will disable "Send" button when private message box is empty. In addition, a tooltip will appear on mouseover to make it clear why the button is disabled.

In addition, placeholder text will prompt the user to enter a message when the private message text box is empty.

rdhammond committed:
Adding an extra check to disable "send" for private messages when textbox
is empty.

rdhammond committed:
Added a tooltip to make it clear to the user to type something in!

rdhammond committed:
Tweaked copy on "enter a message" text for private message button popup.
Added placeholder text for empty private message box.

crookedneighbor commented on website/views/shared/modals/members.jade:
```diff
@@ -46,9 +46,12 @@ script(type='text/ng-template', id='modals/private-message.html')
-    button.btn.btn-primary(ng-click='sendPrivateMessage(profile._id, _message)')=env.t("send")
+    //- Due to a quirk in Bootstrap UI, we need to wrap this so the tooltip can
+    //- appear when disabled.
+    div(style='display:inline-block',tooltip="{{_message ? '' : env.t('needsText')}}")
```
I'm going to trust you on the need to wrap it. Is there a reason it needs to be `display: inline-block` instead of `inline`? If it can just be inline, you could change it to a span and remove the inline style. If it must be `inline-block`, I'd rather you [use the bootstrap class](http://getbootstrap.com/css/#responsive-utilities-classes) for doing inline styles:

```
.visible-xs-inline-block(tooltip="{{_message ? '' : env.t('needsText')}}")
```

rdhammond replied:
I did some research, and using a wrapper div with the tooltip is actually the recommended way of handling our situation in the Bootstrap 3 documents themselves. Link embedded into the comments for clarity/explanation, since it's still a little on the hacky side. Still, Bootstrap themselves recommend it, sooooo.

I tried using .visible-xs-inline-block, and it just ended up displaying as a normal block element. It's possible I'm using it incorrectly, though. In the meantime, I created a compromise of putting "display: inline-block" in a separate class inside the CSS. That way, it's at least out of the style attribute, and we can reuse it if we ever need to do this again :)

crookedneighbor commented on website/views/shared/modals/members.jade:
```diff
@@ -46,9 +46,12 @@ script(type='text/ng-template', id='modals/private-message.html')
     h4=env.t('pmHeading', {name: "{{profile.profile.name}}"})
   .modal-body
-    textarea.form-control(type='text',rows='5',ui-keydown='{"meta-enter":"sendPrivateMessage(profile._id, _message)"}',ng-model='_message')
+    textarea.form-control(type='text',rows='5',ui-keydown='{"meta-enter":"sendPrivateMessage(profile._id, _message)"}',placeholder=env.t('needsTextPlaceholder'),ng-model='_message')
```
It's currently possible to send the message by pressing `meta-enter`, which is `ctrl+enter` on a pc and `cmd+enter` on a mac. You may need to edit the `sendPrivateMessage` function to not send if `_message` is blank.

rdhammond replied:
Since we're dependent on an AJAX call to do the commit and close the form, I just stuck a quick truthy check followed by a return statement. If that's not robust enough, we can do the full null/undefined/empty string check.

Side note: I hit Ctrl-Enter on Chrome and nothing happened, even when the text box had text. I confirmed the event wasn't even firing (and was working when the button was clicked), so it wasn't the new code blocking it. Maybe a ticket that needs to be opened elsewhere? :shrug:

crookedneighbor replied:
@rdhammond Just want to double check two things:

1) You're on a windows machine using Chrome, right? (On a mac, you'd do `cmd + enter`)
2) Does the `ctrl + enter` just not work on this branch? Or does it not work on the website as well?

I don't have access to a PC right now. Can another dev confirm the behavior of `ctrl + enter` when sending PMs?

Alys replied:
Ctrl-Enter on Windows 7 does NOT work to send PMs on the production site (Cmd-Enter on a Mac does work).

I'll try to get this PR reviewed and hopefully merged this weekend.

rdhammond committed:
Documenting the (Bootstrap recommended!) wrapper workaround for tooltips
on disabled elements. Added code to prevent user from posting blank
messages by using meta-enter as a workaround.

Alys commented:
@rdhammond This is a nice little fix! Thank you!

Please tell me your user ID (Settings -> API) so that I can award you a contributor tier.

rdhammond commented:
Woo! ee314a20-4e76-4f36-8fba-10828eb3770f

On Sun, Apr 5, 2015 at 9:38 AM Alys notifications@github.com wrote:

> @rdhammond https://github.com/rdhammond This is a nice little fix!
> Thank you!
> 
> Please tell me your user ID (Settings -> API) so that I can award you a
> contributor tier.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/4910#issuecomment-89771410.

Alys commented:
Tier 2 awarded!

Note that the higher tiers require more effort than the lower ones, so if you make any small changes from now on, you might find that they won't grant you an extra tier immediately, but they will be noted as partial credit towards a tier. There's a bit of information about that at [Contributor Rewards](http://habitrpg.wikia.com/wiki/Contributor_Rewards) but do ask if you have questions.
